### ISSUE TITLE
GPU mask mode problems with OSX

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1459]
All work fine using CPU device. But when using GPU (Intel HD4000 or nvidia GT650M) I get problems.

Here's Intel HD4000. Perhaps you are not honoring device limits?
```
$ ../run/john -test -form:raw-md4-opencl 
Device 1: HD Graphics 4000
Benchmarking: Raw-MD4-opencl [MD4 OpenCL]... OpenCL error (CL_INVALID_BUFFER_SIZE) in file (opencl_rawmd4_fmt_plug.c) at line (171) - (Error creating page-locked memory pinned_saved_keys.)

$ ../run/john -test -form:raw-md5-opencl 
Device 1: HD Graphics 4000
Benchmarking: Raw-MD5-opencl [MD5 OpenCL]... OpenCL error (CL_INVALID_BUFFER_SIZE) in file (opencl_rawmd5_fmt_plug.c) at line (168) - (Error creating page-locked memory pinned_saved_keys.)

$ ../run/john -test -form:raw-sha1-opencl 
Device 1: HD Graphics 4000
Benchmarking: Raw-SHA1-opencl [SHA1 OpenCL]... OpenCL error (CL_INVALID_BUFFER_SIZE) in file (opencl_rawsha1_fmt_plug.c) at line (167) - (Error creating page-locked memory pinned_saved_keys.)
```

And here's the nvidia:
```
$ ../run/john -test -form:raw-md4-opencl -dev=2
Device 2: GeForce GT 650M
Benchmarking: Raw-MD4-opencl [MD4 OpenCL]... Error, crypt_all kernel.

$ ../run/john -test -form:raw-md5-opencl -dev=2
Device 2: GeForce GT 650M
Benchmarking: Raw-MD5-opencl [MD5 OpenCL]... OpenCL error (CL_INVALID_VALUE) in file (opencl_rawmd5_fmt_plug.c) at line (756) - (failed in reading back num cracked hashes.)

$ ../run/john -test -form:raw-sha1-opencl -dev=2
Device 2: GeForce GT 650M
Benchmarking: Raw-SHA1-opencl [SHA1 OpenCL]... Error, crypt_all kernel.
```
Note: The two different problems seen above can both happen with any of the three formats. You either get the "failed in reading back num cracked hashes" or the "Error, crypt_all kernel".

I'd be happy to help but I'm not sure what to look for. The OSX drivers are very good nowadays, I do not think this is a driver bug.

**Comment by magnumripper** [comment]
BTW you should probably fix #1455 first, just in case some of the warnings actually point to a real problem.

**Comment by Sayantan2048** [comment]
I guess openCL 1.1 is being used for Nv devices. In that case I suspect the code block at line 244 in md4_kernel.cl is not working. Even if the code block is working, single work item initializations are problematic and in this case it's bad driver. This is why I tend to avoid them using clFillBuffer on the host side, available only with opencl 1.2 or above.

**Comment by Sayantan2048** [comment]
I'll fix rest of the pinned memory issues.

**Comment by magnumripper** [comment]
Hm, the device list says "OpenCL 1.2" but maybe that's a silly "version name clash" and it's actually merely version 1.2 of the *driver* and has nothing to do with general OpenCL version.

I added a #warning notice and the code in line 244 is indeed compiled - but apparently it doesn't help.

**Comment by magnumripper** [comment]
Scrap that, I screwed up. The device is indeed OpenCL 1.2 and that code for <1.2 is not used.

**Comment by magnumripper** [comment]
...and if I change it to `#if 1`, it works fine.

**Comment by magnumripper** [comment]
So here is a workaround
```diff
@@ -241,7 +241,7 @@ __kernel void md4(__global uint *keys,
        uint len = base & 63;
        uint hash[4];
 
-#if __OPENCL_VERSION__ < 120
+#if __OPENCL_VERSION__ < 120 || APPLE
        if (!gid) {
                out_hash_ids[0] = 0;
                for (i = 0; i < HASH_TABLE_SIZE/32 + 1; i++)
```

**Comment by magnumripper** [comment]
OTOH this seem to be a driver bug only with the nvidia (not with CPU device nor HD4000), so that workaround may blow performance for AMD devices on OSX. I can't test that.

**Comment by Sayantan2048** [comment]
For nvidia can you test 'get_device_version(gpu_id) >= 120' amounts to true or false at line 727?

**Comment by magnumripper** [comment]
Yes, I put a debug print in there - that condition is true and the fillbuffer stuff is executed.

**Comment by magnumripper** [comment]
Heh, the problem is cl_createobj() is called with a kpc of 0 (zero). I'm looking into why.

**Comment by magnumripper** [comment]
HD4000 problem is a driver bug. End time for profiling of FillBuffer stuff was 0. Workaround in aadb6c1.

**Comment by magnumripper** [comment]
OK so back to the nvidia problem. For the record, the driver version is "10.2.7 310.41.25f01". I believe this too is a driver bug. Perhaps we should just force OpenCL 1.1 behavior for nvidias on OSX.

**Comment by magnumripper** [comment]
Here's a verbose run that doesn't end with "Error: crypt_all kernel"
```
$ ../run/john -test -form:raw-md4-opencl -dev=2 -verb=5
initUnicode(UNICODE, ASCII/ASCII)
ASCII -> ASCII -> ASCII
Device 2: GeForce GT 650M
Benchmarking: Raw-MD4-opencl [MD4 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DAPPLE -DDEVICE_INFO=18 -D_OPENCL_COMPILER -DDEV_VER_MAJOR=10 -DDEV_VER_MINOR=2 -D OFFSET_TABLE_SIZE=13 -D HASH_TABLE_SIZE=15 -D SHIFT64_OT_SZ=3 -D SHIFT64_HT_SZ=1 -D NUM_LOADED_HASHES=14 -D NUM_INT_KEYS=1 -D SELECT_CMP_STEPS=4 -D BITMAP_SIZE_BITS_LESS_ONE=32767 -D USE_LOCAL_BITMAPS=1 -D IS_STATIC_GPU_MASK=0 -D LOC_0=-1 -D LOC_1=-1 -D LOC_2=-1 -D LOC_3=-1
Calculating best global worksize (GWS); max. 1s total for crypt_all()
Raw speed figures including buffer transfers:
pass xfer: 11.072us, index xfer: 8.384us, key_loc xfer: 7.488us, fill_buffer_0: 11.744us, fill_buffer_1: 13.440us, crypt: 53.856us, num_cracked xfer: 6.368us
gws:      1024	   9114212 c/s     9114212 rounds/s 112.352us per crypt_all()!
pass xfer: 16.544us, index xfer: 10.752us, key_loc xfer: 10.240us, fill_buffer_0: 11.744us, fill_buffer_1: 9.216us, crypt: 30.176us, num_cracked xfer: 5.984us
gws:      2048	  21636240 c/s    21636240 rounds/s  94.656us per crypt_all()!
pass xfer: 14.816us, index xfer: 8.608us, key_loc xfer: 8.576us, fill_buffer_0: 7.488us, fill_buffer_1: 5.664us, crypt: 48.736us, num_cracked xfer: 4.544us
gws:      4096	  41612483 c/s    41612483 rounds/s  98.432us per crypt_all()+
pass xfer: 24.704us, index xfer: 13.312us, key_loc xfer: 13.728us, fill_buffer_0: 7.264us, fill_buffer_1: 5.504us, crypt: 91.296us, num_cracked xfer: 4.352us
gws:      8192	  51148851 c/s    51148851 rounds/s 160.160us per crypt_all()+
pass xfer: 51.552us, index xfer: 23.904us, key_loc xfer: 23.776us, fill_buffer_0: 7.360us, fill_buffer_1: 5.632us, crypt: 174.496us, num_cracked xfer: 5.888us
gws:     16384	  55993000 c/s    55993000 rounds/s 292.608us per crypt_all()+
pass xfer: 85.696us, index xfer: 50.784us, key_loc xfer: 44.192us, fill_buffer_0: 7.232us, fill_buffer_1: 5.760us, crypt: 342.816us, num_cracked xfer: 4.800us
gws:     32768	  60537984 c/s    60537984 rounds/s 541.280us per crypt_all()+
pass xfer: 177.216us, index xfer: 83.968us, key_loc xfer: 90.848us, fill_buffer_0: 9.216us, fill_buffer_1: 5.664us, crypt: 655.616us, num_cracked xfer: 4.704us
gws:     65536	  63798635 c/s    63798635 rounds/s   1.027ms per crypt_all()+
pass xfer: 347.328us, index xfer: 169.184us, key_loc xfer: 162.816us, fill_buffer_0: 7.040us, fill_buffer_1: 5.696us, crypt: 1.287ms, num_cracked xfer: 4.256us
gws:    131072	  66055992 c/s    66055992 rounds/s   1.984ms per crypt_all()+
pass xfer: 659.104us, index xfer: 330.464us, key_loc xfer: 321.568us, fill_buffer_0: 7.040us, fill_buffer_1: 5.696us, crypt: 2.578ms, num_cracked xfer: 4.992us
gws:    262144	  67081559 c/s    67081559 rounds/s   3.907ms per crypt_all()+
pass xfer: 1.380ms, index xfer: 672.544us, key_loc xfer: 672.544us, fill_buffer_0: 7.040us, fill_buffer_1: 5.696us, crypt: 5.136ms, num_cracked xfer: 5.088us
gws:    524288	  66538603 c/s    66538603 rounds/s   7.879ms per crypt_all()
pass xfer: 2.760ms, index xfer: 1.344ms, key_loc xfer: 1.347ms, fill_buffer_0: 7.040us, fill_buffer_1: 5.696us, crypt: 10.265ms, num_cracked xfer: 4.960us
gws:   1048576	  66639550 c/s    66639550 rounds/s  15.735ms per crypt_all()
pass xfer: 5.473ms, index xfer: 2.683ms, key_loc xfer: 2.680ms, fill_buffer_0: 7.072us, fill_buffer_1: 5.696us, crypt: 20.715ms, num_cracked xfer: 4.992us
gws:   2097152	  66427523 c/s    66427523 rounds/s  31.570ms per crypt_all()
pass xfer: 10.897ms, index xfer: 5.274ms, key_loc xfer: 5.287ms, fill_buffer_0: 6.976us, fill_buffer_1: 5.952us, crypt: 41.379ms, num_cracked xfer: 4.640us
gws:   4194304	  66728912 c/s    66728912 rounds/s  62.855ms per crypt_all()
Hardware resources exhausted
Max local worksize 1024, Local worksize (LWS) 256, global worksize (GWS) 262144
OpenCL error (CL_INVALID_VALUE) in file (opencl_rawmd4_fmt_plug.c) at line (769) - (failed in reading back num cracked hashes.)
```

And here's a run that happens to end with "Error: crypt_all kernel"
```
$ ../run/john -test -form:raw-md4-opencl -dev=2 -verb=5
initUnicode(UNICODE, ASCII/ASCII)
ASCII -> ASCII -> ASCII
Device 2: GeForce GT 650M
Benchmarking: Raw-MD4-opencl [MD4 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DAPPLE -DDEVICE_INFO=18 -D_OPENCL_COMPILER -DDEV_VER_MAJOR=10 -DDEV_VER_MINOR=2 -D OFFSET_TABLE_SIZE=13 -D HASH_TABLE_SIZE=15 -D SHIFT64_OT_SZ=3 -D SHIFT64_HT_SZ=1 -D NUM_LOADED_HASHES=14 -D NUM_INT_KEYS=1 -D SELECT_CMP_STEPS=4 -D BITMAP_SIZE_BITS_LESS_ONE=32767 -D USE_LOCAL_BITMAPS=1 -D IS_STATIC_GPU_MASK=0 -D LOC_0=-1 -D LOC_1=-1 -D LOC_2=-1 -D LOC_3=-1
Calculating best global worksize (GWS); max. 1s total for crypt_all()
Raw speed figures including buffer transfers:
pass xfer: 10.688us, index xfer: 8.608us, key_loc xfer: 7.776us, fill_buffer_0: 11.808us, fill_buffer_1: 13.184us, crypt: 53.376us, num_cracked xfer: 7.008us
gws:      1024	   9106431 c/s     9106431 rounds/s 112.448us per crypt_all()!
pass xfer: 9.536us, index xfer: 6.496us, key_loc xfer: 6.240us, fill_buffer_0: 7.264us, fill_buffer_1: 5.952us, crypt: 31.200us, num_cracked xfer: 5.376us
gws:      2048	  28419182 c/s    28419182 rounds/s  72.064us per crypt_all()!
pass xfer: 14.496us, index xfer: 8.640us, key_loc xfer: 8.544us, fill_buffer_0: 7.296us, fill_buffer_1: 5.632us, crypt: 48.672us, num_cracked xfer: 5.856us
gws:      4096	  41316978 c/s    41316978 rounds/s  99.136us per crypt_all()+
pass xfer: 24.608us, index xfer: 13.408us, key_loc xfer: 13.408us, fill_buffer_0: 7.296us, fill_buffer_1: 5.280us, crypt: 90.208us, num_cracked xfer: 4.448us
gws:      8192	  51633723 c/s    51633723 rounds/s 158.656us per crypt_all()+
pass xfer: 48.960us, index xfer: 23.904us, key_loc xfer: 23.392us, fill_buffer_0: 7.296us, fill_buffer_1: 5.600us, crypt: 176.640us, num_cracked xfer: 5.664us
gws:     16384	  56214317 c/s    56214317 rounds/s 291.456us per crypt_all()+
pass xfer: 87.680us, index xfer: 43.648us, key_loc xfer: 44.128us, fill_buffer_0: 7.296us, fill_buffer_1: 5.568us, crypt: 345.440us, num_cracked xfer: 6.432us
gws:     32768	  60659913 c/s    60659913 rounds/s 540.192us per crypt_all()+
pass xfer: 174.944us, index xfer: 83.872us, key_loc xfer: 84us, fill_buffer_0: 7.264us, fill_buffer_1: 5.984us, crypt: 684.832us, num_cracked xfer: 4.352us
gws:     65536	  62698995 c/s    62698995 rounds/s   1.045ms per crypt_all()+
pass xfer: 342.368us, index xfer: 184.096us, key_loc xfer: 169.728us, fill_buffer_0: 7.072us, fill_buffer_1: 5.760us, crypt: 1.374ms, num_cracked xfer: 4.800us
gws:    131072	  62777794 c/s    62777794 rounds/s   2.087ms per crypt_all()
pass xfer: 657.984us, index xfer: 322.112us, key_loc xfer: 322.016us, fill_buffer_0: 7.264us, fill_buffer_1: 5.888us, crypt: 5.663ms, num_cracked xfer: 7.904us
gws:    262144	  37518262 c/s    37518262 rounds/s   6.987ms per crypt_all()
pass xfer: 2.650ms, index xfer: 1.325ms, key_loc xfer: 1.296ms, fill_buffer_0: 11.872us, fill_buffer_1: 9.632us, crypt: 5.388ms, num_cracked xfer: 5.152us
gws:    524288	  49056242 c/s    49056242 rounds/s  10.687ms per crypt_all()
pass xfer: 2.757ms, index xfer: 1.297ms, key_loc xfer: 1.349ms, fill_buffer_0: 7.360us, fill_buffer_1: 5.952us, crypt: 10.963ms, num_cracked xfer: 4.640us
gws:   1048576	  63991376 c/s    63991376 rounds/s  16.386ms per crypt_all()+
pass xfer: 5.415ms, index xfer: 2.651ms, key_loc xfer: 2.663ms, fill_buffer_0: 6.912us, fill_buffer_1: 5.728us, crypt: 21.908ms, num_cracked xfer: 6.176us
gws:   2097152	  64214529 c/s    64214529 rounds/s  32.658ms per crypt_all()
pass xfer: 21.538ms, index xfer: 10.475ms, key_loc xfer: 5.307ms, fill_buffer_0: 7.328us, fill_buffer_1: 6.016us, crypt: 44.056ms, num_cracked xfer: 4.960us
gws:   4194304	  51529182 c/s    51529182 rounds/s  81.396ms per crypt_all()
Hardware resources exhausted
Max local worksize 1024, Error, crypt_all kernel.
```
Note that it seems to fail when auto-tuning LWS. I tried pegging it to 1024 but only got the `OpenCL error (CL_INVALID_VALUE) in file (opencl_rawmd4_fmt_plug.c) at line (769) - (failed in reading back num cracked hashes.)`.

**Comment by magnumripper** [comment]
Here's overriding auto-tune (no real difference)
```
$ LWS=8 GWS=64 ../run/john -test -form:raw-md4-opencl -dev=2 -verb=5
initUnicode(UNICODE, ASCII/ASCII)
ASCII -> ASCII -> ASCII
Device 2: GeForce GT 650M
Benchmarking: Raw-MD4-opencl [MD4 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DAPPLE -DDEVICE_INFO=18 -D_OPENCL_COMPILER -DDEV_VER_MAJOR=10 -DDEV_VER_MINOR=2 -D OFFSET_TABLE_SIZE=20 -D HASH_TABLE_SIZE=17 -D SHIFT64_OT_SZ=16 -D SHIFT64_HT_SZ=1 -D NUM_LOADED_HASHES=14 -D NUM_INT_KEYS=1 -D SELECT_CMP_STEPS=4 -D BITMAP_SIZE_BITS_LESS_ONE=32767 -D USE_LOCAL_BITMAPS=1 -D IS_STATIC_GPU_MASK=0 -D LOC_0=-1 -D LOC_1=-1 -D LOC_2=-1 -D LOC_3=-1
Local worksize (LWS) 8, global worksize (GWS) 64
OpenCL error (CL_INVALID_VALUE) in file (opencl_rawmd4_fmt_plug.c) at line (769) - (failed in reading back num cracked hashes.)
```

**Comment by magnumripper** [comment]
Hmm here's some more info
```
$ CL_LOG_ERRORS=stderr ../run/john -test -form:raw-md4-opencl -dev=2 -verb=5
(...)
pass xfer: 656.864us, index xfer: 321.568us, key_loc xfer: 321.600us, fill_buffer_0: 6.496us, fill_buffer_1: 5.088us, crypt: 2.383ms, num_cracked xfer: 3.968us
gws:    262144	  70871182 c/s    70871182 rounds/s   3.698ms per crypt_all()+
pass xfer: 1.316ms, index xfer: 676.160us, key_loc xfer: 644.576us, fill_buffer_0: 6.400us, fill_buffer_1: 5.184us, crypt: 5.029ms, num_cracked xfer: 4.352us
gws:    524288	  68246192 c/s    68246192 rounds/s   7.682ms per crypt_all()
pass xfer: 2.722ms, index xfer: 1.275ms, key_loc xfer: 1.304ms, fill_buffer_0: 6.304us, fill_buffer_1: 5.088us, crypt: 9.965ms, num_cracked xfer: 4.640us
gws:   1048576	  68604687 c/s    68604687 rounds/s  15.284ms per crypt_all()
pass xfer: 5.490ms, index xfer: 2.662ms, key_loc xfer: 2.682ms, fill_buffer_0: 6.432us, fill_buffer_1: 5.376us, crypt: 19.845ms, num_cracked xfer: 4.672us
gws:   2097152	  68313774 c/s    68313774 rounds/s  30.698ms per crypt_all()
pass xfer: 10.874ms, index xfer: 5.253ms, key_loc xfer: 5.346ms, fill_buffer_0: 6.304us, fill_buffer_1: 4.768us, crypt: 39.470ms, num_cracked xfer: 4.384us
gws:   4194304	  68803043 c/s    68803043 rounds/s  60.961ms per crypt_all()
Hardware resources exhausted
Max local worksize 1024, Local worksize (LWS) 128, global worksize (GWS) 262144
[CL_INVALID_COMMAND_QUEUE] : OpenCL Fatal Error : Finish returned an error

Break on OpenCLFatalBreak to debug.
[CL_INVALID_VALUE] : OpenCL Error : Driver call for CL_COMMAND_READ_BUFFER error code 0xffffffe2 (-30)
Break on OpenCLErrorBreak to debug.
[CL_INVALID_OPERATION] : OpenCL Error : Some previous asynchronously enqueued event on this queue retured an error
Break on OpenCLErrorBreak to debug.
[CL_INVALID_VALUE] : OpenCL Error : clEnqueueReadBuffer failed: Invalid value
Break on OpenCLErrorBreak to debug.
OpenCL error (CL_INVALID_VALUE) in file (opencl_rawmd4_fmt_plug.c) at line (769) - (failed in reading back num cracked hashes.)
```

**Comment by magnumripper** [comment]
On a side note, the `(get_device_version(gpu_id)` function call is far too slow to be used within crypt_all. You should run it once in init() or something, and cache the result in a variable.

**Comment by magnumripper** [comment]
Fixed that in 2eee3dc

**Comment by magnumripper** [comment]
I added a workaround for now, forcing OpenCL 1.1 behavior for nvidias on Apple platform.

**Comment by magnumripper** [comment]
OK, now I see an other problem. This only applies to mask mode.
```
$ echo bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb > test.in
$ ../run/john -form:raw-md4-opencl -dev=2 test.in -mask:?l?l?l?l?l?l
Device 2: GeForce GT 650M
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD4-opencl [MD4 OpenCL])
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 8192
Press 'q' or Ctrl-C to abort, almost any other key for status
OpenCL error (CL_OUT_OF_RESOURCES) in file (opencl_rawmd4_fmt_plug.c) at line (773) - (failed in clEnqueueNDRangeKernel)

$ ../run/john -form:raw-md5-opencl -dev=2 test.in -mask:?l?l?l?l?l?l
Device 2: GeForce GT 650M
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5-opencl [MD5 OpenCL])
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 4096
Press 'q' or Ctrl-C to abort, almost any other key for status
OpenCL error (CL_OUT_OF_RESOURCES) in file (opencl_rawmd5_fmt_plug.c) at line (770) - (failed in clEnqueueNDRangeKernel)

$ echo bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb > test.in
$ ../run/john -form:raw-sha1-opencl -dev=2 test.in -mask:?l?l?l?l?l?l
Device 2: GeForce GT 650M
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-SHA1-opencl [SHA1 OpenCL])
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 32768
Press 'q' or Ctrl-C to abort, almost any other key for status
OpenCL error (CL_OUT_OF_RESOURCES) in file (opencl_rawsha1_fmt_plug.c) at line (791) - (failed in clEnqueueNDRangeKernel)
```
Before your work with new hash checking, this was not a problem.

**Comment by magnumripper** [comment]
Apparently mask_int_cand.num_int_cand end up as 17,576 (26^3) despite mask_int_cand_target is 10,000. If I use ?d?d?d?d?d?d?d?d?d?d?d it ends up as 10,000 and format works fine.

**Comment by Sayantan2048** [comment]
Fixed 585021b41bca8b218f

**Comment by magnumripper** [comment]
Works fine now!
